community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/runtime/commands/expressions/FilterFunction.scala

```diff
@@ -21,15 +21,22 @@ package org.neo4j.cypher.internal.compatibility.v3_3.runtime.commands.expression
 
 import org.neo4j.cypher.internal.compatibility.v3_3.runtime.ExecutionContext
 import org.neo4j.cypher.internal.compatibility.v3_3.runtime.commands.predicates.Predicate
+import org.neo4j.cypher.internal.compatibility.v3_3.runtime.helpers.ListSupport
 import org.neo4j.cypher.internal.compatibility.v3_3.runtime.pipes.QueryState
-import org.neo4j.cypher.internal.compiler.v3_3.helpers.ListSupport
+import org.neo4j.values.AnyValue
+import org.neo4j.values.virtual.VirtualValues
 
 case class FilterFunction(collection: Expression, id: String, predicate: Predicate)
   extends NullInNullOutExpression(collection)
   with ListSupport
   with Closure {
-  def compute(value: Any, m: ExecutionContext)(implicit state: QueryState) =
-    makeTraversable(value).filter(element => predicate.isTrue(m.newWith1(id, element)  ))
+
+  def compute(value: AnyValue, m: ExecutionContext)(implicit state: QueryState) = {
+    val traversable = makeTraversable(value)
+    VirtualValues.filter(traversable, new java.util.function.Function[AnyValue, java.lang.Boolean]() {
```

Comment:
Yeah the main target is java so here scala will have to suffer a bit. I think when we upgrade to newer version of scala it will handle java lambdas better. We could of course add some utility but that will also add indirection, I think I prefer faster over prettier in the runtime.